OpenRA.Mods.Common/Graphics/ContrailRenderable.cs

```diff
@@ -92,54 +94,82 @@ public void Render(WorldRenderer wr)
 			var screenWidth = wr.ScreenVector(new WVec(1, 0, 0))[0];
 			var wcr = Game.Renderer.WorldRgbaColorRenderer;
 
-			var startColor = this.startColor;
-			if (usePlayerStartColor)
-				startColor = Color.FromArgb(this.startColor.A, owner.OwnerColor());
+			var startColor = usePlayerStartColor ? Color.FromArgb(this.startColor.A, owner.OwnerColor()) : this.startColor;
+			var endColor = usePlayerEndColor ? Color.FromArgb(this.endColor.A, owner.OwnerColor()) : this.endColor;
 
-			var endColor = this.endColor;
-			if (usePlayerEndColor)
-				endColor = Color.FromArgb(this.endColor.A, owner.OwnerColor());
-
-			// Start of the first line segment is the tail of the list - don't smooth it.
 			var curPos = trail[Index(next - skip - 1)];
 			var curColor = startColor;
+			var curWidth = startWidth.Length * screenWidth;
+			var endr = float3.Zero;
+			var endl = float3.Zero;
 
 			for (var i = 1; i < renderLength; i++)
 			{
+				var pos = curPos;
```

Comment:
Couldn't the renderer like the trail also cache the quads of the last x positions.

Seems expensive to recalculate the color and positions for all segements each time.

But i see you considering Fog below. So likely no option.

Reply:
indeed